### PR TITLE
feat(space): reason-aware blocked task banner with gate approval UI

### PR DESCRIPTION
**Pull request by lsm**
## Summary
- Replace generic amber blocked banner with `TaskBlockedBanner` component that shows distinct UI per `blockReason` (human_input_requested, gate_rejected, execution_failed, agent_crashed, dependency_failed, workflow_invalid)
- Wire gate approval flow: `gate_rejected` banner has "Review & Approve" button that opens `GateArtifactsView` inline with full diff review + approve/reject
- Add `listGateData` and `approveGate` methods to `spaceStore` (RPCs already existed, just needed frontend wiring)

Gap #2b from the space autonomy gap analysis.

## Test plan
- [x] 13 new unit tests for TaskBlockedBanner covering all 6 block reasons, gate review flow, resume actions
- [x] Updated SpaceTaskPane test for new always-show-banner behavior
- [x] Full web test suite passes (6953/6954, 1 pre-existing failure unrelated)
- [x] Typecheck, lint, knip all clean
- [x] E2E `space-task-status-control` test compatible (same `task-blocked-banner` testid preserved)

### CODE REVIEW

File: packages/web/src/lib/space-store.ts [hunk 1]
```diff
@@ -1642,6 +1642,52 @@ class SpaceStore {
...
+				updatedAt: number;
+			}>;
+		}>('spaceWorkflowRun.listGateData', { runId });
+		return result?.gateData ?? [];
+	}
+
+	/**
+	 * Approve or reject a gate on a workflow run.
+	 */
+	async approveGate(
```
Comment (lsm) [inline packages/web/src/lib/space-store.ts]: **Dead code — `approveGate` is never called from this PR.**

`GateArtifactsView` (line 119) and `ReadOnlyWorkflowCanvas` (line 125) both call `hub.request('spaceWorkflowRun.approveGate', ...)` directly via `connectionManager`, bypassing `spaceStore` entirely. This method has zero callers and no tests.

Options:
- Refactor `GateArtifactsView` to call `spaceStore.approveGate()` (centralises the RPC call), **or**
- Remove this method and add it in a follow-up PR when it is actually wired up.

File: packages/web/src/components/space/TaskBlockedBanner.tsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	useEffect(() => {
+		if (reason !== 'gate_rejected' || !task.workflowRunId) return;
+
+		setGateLoading(true);
+		spaceStore
+			.listGateData(task.workflowRunId)
+			.then((records) => {
+				// Find the rejected/unapproved gate
+				const rejected = records.find(
+					(r) => r.data?.approved === false || r.data?.waiting === true
```
Comment (lsm) [inline packages/web/src/components/space/TaskBlockedBanner.tsx:114]: **Ambiguous gate selection in multi-gate workflows.**

`find()` returns the *first* record ordered alphabetically by `gateId`. A workflow graph may have multiple human-approval gates; the one that blocked this task might not be first in the list.

Suggested options:
1. Store a `blockingGateId` on `SpaceTask` when it transitions to `blocked` — cleanest long-term fix.
2. Filter by gate node ID embedded in the task's workflow metadata if available.
3. At minimum add a comment explaining the current limitation.

File: packages/web/src/components/space/TaskBlockedBanner.tsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+				);
+				if (rejected) {
+					setPendingGate({ gateId: rejected.gateId, data: rejected.data });
+				}
+			})
+			.catch(() => {
+				// Gate data fetch is best-effort
+			})
+			.finally(() => setGateLoading(false));
+	}, [reason, task.workflowRunId]);
```
Comment (lsm) [inline packages/web/src/components/space/TaskBlockedBanner.tsx:129]: **Missing cleanup — can update state on an unmounted component.**

If `TaskBlockedBanner` unmounts before `listGateData` resolves (e.g. user switches tasks quickly), the `.then()` / `.finally()` callbacks still fire.

```ts
useEffect(() => {
  if (reason !== 'gate_rejected' || !task.workflowRunId) return;
  let cancelled = false;
  setGateLoading(true);
  spaceStore
    .listGateData(task.workflowRunId)
    .then((records) => {
      if (cancelled) return;
      const rejected = records.find(
        (r) => r.data?.approved === false || r.data?.waiting === true
      );
      if (rejected) setPendingGate({ gateId: rejected.gateId, data: rejected.data });
    })
    .catch(() => {})
    .finally(() => { if (!cancelled) setGateLoading(false); });
  return () => { cancelled = true; };
}, [reason, task.workflowRunId]);
```

File: packages/web/src/components/space/TaskBlockedBanner.tsx [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	icon: '⚠',
+};
+
+export function TaskBlockedBanner({ task, spaceId, onStatusTransition }: TaskBlockedBannerProps) {
+	const reason = task.blockReason as SpaceBlockReason | null;
+	const config = (reason && REASON_CONFIG[reason]) || FALLBACK_CONFIG;
+
+	const [showGateReview, setShowGateReview] = useState(false);
+	const [pendingGate, setPendingGate] = useState<PendingGate | null>(null);
+	const [gateLoading, setGateLoading] = useState(false);
```
Comment (lsm) [inline packages/web/src/components/space/TaskBlockedBanner.tsx]: **One-tick "Resume" button flash for `gate_rejected` tasks.**

`gateLoading` starts `false` and `pendingGate` is `null`, so the condition `!pendingGate && !gateLoading` is immediately true — the Resume button renders for one tick before `setGateLoading(true)` fires.

Fix:
```ts
const [gateLoading, setGateLoading] = useState(
  reason === 'gate_rejected' && !!task.workflowRunId
);
```

File: packages/web/src/components/space/__tests__/TaskBlockedBanner.test.tsx [hunk 5]
```diff
@@ -0,0 +1,258 @@
+// @ts-nocheck
```
Comment (lsm) [inline packages/web/src/components/space/__tests__/TaskBlockedBanner.test.tsx]: **`// @ts-nocheck` suppresses all TypeScript errors in this file.** No other test file in this repo uses it. The JSX-after-mock pattern is fine without it — Vitest hoists `vi.mock()` calls correctly. Please remove it and fix any type errors that surface (likely the `makeTask` helper is missing a few required `SpaceTask` fields that `@ts-nocheck` was masking).

**Review by lsm — COMMENTED** [review]
## Review: `feat(space): reason-aware blocked task banner with gate approval UI`

Good direction overall — the per-reason UX is a clear improvement over the generic amber banner, and the 13 unit tests cover the happy paths well. A few issues need addressing before merge.

### ❌ Blocking

1. **`approveGate` is dead code** (see inline on `space-store.ts`). `GateArtifactsView` already calls `hub.request('spaceWorkflowRun.approveGate', ...)` directly via `connectionManager`. The new `spaceStore.approveGate()` is never invoked anywhere. Either wire `GateArtifactsView` to use `spaceStore.approveGate()` for consistency, or drop it from this PR.

2. **Ambiguous gate selection in multi-gate workflows** (see inline on `TaskBlockedBanner.tsx:105`). `records.find(r => r.data?.approved === false || r.data?.waiting === true)` picks the *first alphabetically-ordered* gate that matches. A workflow may have multiple approval gates; the blocking one might not be first. At a minimum, document the limitation, or add a `blockingGateId` field to `SpaceTask`.

### ⚠️ Non-blocking (should fix before merge)

3. **`useEffect` missing cleanup** — no guard against state updates after unmount. Add a `cancelled` flag and return a cleanup function (see inline comment).

4. **Initial `gateLoading = false` flash** — for `gate_rejected` + `workflowRunId`, the "Resume" button renders for one tick before `setGateLoading(true)` fires. Initialize: `useState(reason === 'gate_rejected' && !!task.workflowRunId)`.

5. **`// @ts-nocheck` in test file** — no other test file in this codebase uses it. Remove it and fix any type errors that surface.

### 📝 Minor

6. **Missing edge-case tests**: `gate_rejected` with `workflowRunId: null` (should show Resume immediately), `onDecision` callback path, and the intermediate loading state.

7. **Emoji variation selectors** — `'⚠'` (U+26A0) and `'⛓'` (U+26D3) render as symbol glyphs on some platforms. Use `'⚠️'` and `'⛓️'` for consistent emoji rendering.

File: packages/web/src/components/space/TaskBlockedBanner.tsx [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+	/** Called when the user triggers a status transition (e.g. Resume → in_progress) */
+	onStatusTransition?: (newStatus: SpaceTaskStatus) => void;
+}
+
+interface PendingGate {
+	gateId: string;
+	data: Record<string, unknown>;
+}
+
+const REASON_CONFIG: Record<
```
Comment (lsm) [inline packages/web/src/components/space/TaskBlockedBanner.tsx]: **Nit:** `Record<string, …>` gives no exhaustiveness signal if a new `SpaceBlockReason` variant is added. Consider:
```ts
const REASON_CONFIG: Partial<Record<SpaceBlockReason, { label: string; border: string; bg: string; title: string; icon: string }>> = {
```
The existing `|| FALLBACK_CONFIG` already handles the missing-key case at runtime, so this is purely a type-safety improvement.

File: packages/web/src/components/space/TaskBlockedBanner.tsx [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+const FALLBACK_CONFIG = {
+	label: 'Blocked',
+	border: 'border-amber-500/30',
+	bg: 'bg-amber-500/10',
+	title: 'text-amber-300',
+	icon: '⚠️',
+};
+
+export function TaskBlockedBanner({ task, spaceId, onStatusTransition }: TaskBlockedBannerProps) {
+	const reason = task.blockReason as SpaceBlockReason | null;
```
Comment (lsm) [inline packages/web/src/components/space/TaskBlockedBanner.tsx]: **Nit:** Redundant cast — `task.blockReason` is already typed `SpaceBlockReason | null` on `SpaceTask`. `as SpaceBlockReason | null` is a no-op and can be removed:
```ts
const reason = task.blockReason;
```

File: packages/web/src/components/space/TaskBlockedBanner.tsx [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+						<p class="mt-1 text-xs text-blue-400/70">Reply below to continue</p>
+					)}
+				</div>
+
+				<div class="flex items-center gap-1.5 flex-shrink-0">
+					{reason === 'gate_rejected' && pendingGate && (
+						<button
+							type="button"
+							onClick={() => setShowGateReview(true)}
+							disabled={gateLoading}
```
Comment (lsm) [inline packages/web/src/components/space/TaskBlockedBanner.tsx]: **Nit:** `disabled={gateLoading}` is effectively always `false` here. The button only renders when `pendingGate` is set, and `pendingGate` is only set inside `.then()` — which runs after `.finally()` has already set `gateLoading = false`. You could drop the `disabled` prop, or restructure the loading/pending conditions so it actually takes effect (e.g. keep the button visible during loading but disabled).

**Review by lsm — COMMENTED** [review]
## Follow-up review — all previous issues resolved ✅

Second commit addresses every item from the first review:

| Issue | Status |
|---|---|
| `approveGate` dead code | ✅ Removed |
| Ambiguous gate selection | ✅ Documented with TODO comment |
| `useEffect` missing cleanup | ✅ `cancelled` flag + cleanup return |
| `gateLoading` flash | ✅ Initialized to `true` for `gate_rejected + workflowRunId` |
| `@ts-nocheck` in tests | ✅ Removed |
| Edge-case tests | ✅ 3 new: null workflowRunId, loading state, onDecision |
| Emoji variation selectors | ✅ All icons now use variation selector U+FE0F |

### Remaining nits (non-blocking)

1. **`REASON_CONFIG` typed as `Record<string, …>` rather than `Partial<Record<SpaceBlockReason, …>>`** — if a new `SpaceBlockReason` is added (e.g. `timeout`), TypeScript won't flag the missing entry; it silently falls through to `FALLBACK_CONFIG`. The runtime behaviour is correct (fallback is safe), but `Partial<Record<SpaceBlockReason, …>>` would give compile-time coverage for free.

2. **Redundant cast on line 88** — `task.blockReason as SpaceBlockReason | null` is already the declared type on `SpaceTask.blockReason`; the cast is a no-op.

3. **`disabled={gateLoading}` is effectively always `false`** when the "Review & Approve" button renders. `pendingGate` is only set inside `.then()`, which only fires after `.finally()` sets `gateLoading = false`. So by the time `pendingGate && reason === 'gate_rejected'` is true, `gateLoading` is already `false`. The prop is harmless but dead.

4. **Vestigial mock** — `vi.mock('../../../lib/utils')` in `TaskBlockedBanner.test.tsx` mocks a `cn` helper that `TaskBlockedBanner` never imports. Safe to delete.

**None of these block merge.** Code is correct, all hooks are clean, test coverage is thorough. LGTM.

File: packages/web/src/components/space/__tests__/TaskBlockedBanner.test.tsx [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+
+vi.mock('../../../lib/space-store', () => ({
+	spaceStore: {
+		listGateData: (...args: unknown[]) => mockListGateData(...args),
+	},
+}));
+
+vi.mock('../../../lib/utils', () => ({
+	cn: (...args: unknown[]) => args.filter(Boolean).join(' '),
+}));
```
Comment (lsm) [inline packages/web/src/components/space/__tests__/TaskBlockedBanner.test.tsx]: **Nit:** This mock is unused — `TaskBlockedBanner` doesn't import from `../../../lib/utils`. Safe to delete.

File: packages/web/src/components/space/TaskBlockedBanner.tsx [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+			/>
+		);
+	}
+
+	return (
+		<div
+			class={`mx-4 mt-2 rounded-lg border ${config.border} ${config.bg} px-3 py-2`}
+			data-testid="task-blocked-banner"
+		>
+			<div class="flex items-start justify-between gap-2">
```
Comment (lsm) [inline packages/web/src/components/space/TaskBlockedBanner.tsx:150]: **Cosmetic inconsistency.** This `gate-resume-btn` has no background; the `task-resume-btn` below (line ~163) has `bg-dark-700 hover:bg-dark-600`. Both buttons do the same thing (`onStatusTransition?.('in_progress')`). Consider giving them the same classes, or document why the different visual weight is intentional.

File: packages/web/src/components/space/TaskBlockedBanner.tsx [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+				}
+			})
+			.catch(() => {
+				// Gate data fetch is best-effort
+			})
+			.finally(() => {
+				if (!cancelled) setGateLoading(false);
+			});
+		return () => {
+			cancelled = true;
```
Comment (lsm) [inline packages/web/src/components/space/TaskBlockedBanner.tsx:127]: **Cosmetic / spacing.** Old banner used `mb-2`; this uses `mx-4 mt-2`. The horizontal margins are a nice improvement. Just worth a manual visual check that the gap between the banner and the thread below it is still comfortable — there is no bottom margin here.

**Review by lsm — COMMENTED** [review]
## Whole-PR review (3rd pass — final state, 3 commits)

### What this PR does

Replaces a generic amber `"Blocked"` banner in `SpaceTaskPane` with a new `TaskBlockedBanner` component that renders distinct UI per `blockReason`, and wires up the gate-approval flow (`gate_rejected` → inline `GateArtifactsView`). Adds `spaceStore.listGateData()` to centralise the RPC call. Covers all 6 block reasons.

---

### `TaskBlockedBanner.tsx` — ✅ Clean

**Types / exhaustiveness**
`REASON_CONFIG` is now `Partial<Record<SpaceBlockReason, …>>` — TypeScript will surface a missing entry the moment a new variant is added to the union. All 6 current values are present.

**Hook correctness**
- `gateLoading` initializes to `reason === 'gate_rejected' && !!task.workflowRunId` — correct; prevents the one-tick Resume-button flash for tasks that need a fetch.
- `useEffect` deps `[reason, task.workflowRunId]` are primitive-valued; no stale-closure risk.
- Cleanup (`let cancelled = false; return () => { cancelled = true; }`) is properly guarded in both `.then()` and `.finally()`.
- `.catch(() => {})` swallows fetch errors silently — the "best-effort" comment documents this. The fallback is the "Resume" button, which is a reasonable degraded state.

**Two minor cosmetic observations (non-blocking)**

1. **Resume button styling inconsistency.** `gate-resume-btn` (`gate_rejected` fallback) has no background class; `task-resume-btn` (`execution_failed` / `agent_crashed`) has `bg-dark-700 hover:bg-dark-600`. Both buttons call the same `onStatusTransition?.('in_progress')`. Either give them the same visual treatment or make the styling difference intentional and document why.

2. **Banner margin change.** The old banner used `mb-2`; the new banner uses `mx-4 mt-2`. The horizontal margins keep the banner from touching the pane edges — this looks intentional — but there is no bottom margin, which changes the gap between the banner and the thread below it. Visual change worth a quick manual check.

---

### `space-store.ts` — ✅ Clean

`listGateData` follows the same `connectionManager.getHubIfConnected()` pattern as every other store method. Response shape matches the backend (`{ gateData: [...] }` unwrapped to the inner array, `?? []` fallback). `approveGate` correctly removed — `GateArtifactsView` calls the RPC directly.

---

### `SpaceTaskPane.tsx` — ✅ Clean

`runtimeSpaceId = spaceId ?? task.spaceId` — `task.spaceId` is a required `string` on `SpaceTask`, so `runtimeSpaceId` is always `string`, satisfying `TaskBlockedBannerProps.spaceId: string`. The condition change from `task.status === 'blocked' && task.result` to just `task.status === 'blocked'` is intentional and tested.

---

### `TaskBlockedBanner.test.tsx` — ✅ 16 tests, thorough

| Scenario | Test |
|---|---|
| null blockReason (amber fallback) | ✅ |
| `human_input_requested` — blue, result text, reply hint | ✅ |
| `gate_rejected` + workflowRunId → Review & Approve button | ✅ |
| `gate_rejected` + null workflowRunId → immediate Resume | ✅ |
| `gate_rejected` + empty gate list → Resume fallback | ✅ |
| Loading state — no button flash | ✅ |
| `execution_failed` — red, Resume | ✅ |
| `agent_crashed` — red, Resume | ✅ |
| `dependency_failed` — gray, no buttons | ✅ |
| `workflow_invalid` — red, no buttons | ✅ |
| Resume → `onStatusTransition('in_progress')` | ✅ |
| Open GateArtifactsView → banner hides | ✅ |
| `onClose` → banner restores | ✅ |
| `onDecision` → banner restores | ✅ |
| Result text present / absent | ✅ |

No `@ts-nocheck`, no unused mocks, cleanup in both `beforeEach` and `afterEach`.

---

### `SpaceTaskPane.test.tsx` — ✅

`listGateData` mock added (needed since `TaskBlockedBanner` calls it on mount for gate-blocked tasks). The one inverted assertion correctly flipped to match the new always-show behavior.

---

### Summary

The PR is correct, well-tested, and the code is clean. Both cosmetic observations above are truly minor. **LGTM — ready to merge.**

**Review by lsm — COMMENTED** [review]
4th commit addresses both cosmetic observations from the previous round:

- `mb-2` restored on the banner wrapper — spacing between banner and thread below is back to match the old banner
- `gate-resume-btn` now has identical classes to `task-resume-btn` (`bg-dark-700 hover:bg-dark-600 rounded`) — both Resume buttons look and behave the same

No other changes. The diff is exactly two lines.

**This PR is done. LGTM — ready to merge.** 🟢
